staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1281,6 +1419,71 @@ func (c *cacheWatcher) nextBookmarkTime(now time.Time, bookmarkFrequency time.Du
 	return heartbeatTime, true
 }
 
+// wasBookmarkAfterResourceVersionReceived same as wasBookmarkAfterResourceVersionReceivedLocked just acquires a lock
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionReceived() bool {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+	return c.wasBookmarkAfterResourceVersionReceivedLocked()
+}
+
+// wasBookmarkAfterResourceVersionReceivedLocked checks if the given cacheWatcher have seen a bookmark event >=  bookmarkAfterResourceVersion
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionReceivedLocked() bool {
+	return c.state != cacheWatcherWaitingForBookmark
+}
+
+// markBookmarkAfterResourceVersionAsReceived indicates that the given cacheWatcher have seen a bookmark event >= bookmarkAfterResourceVersion
+func (c *cacheWatcher) markBookmarkAfterResourceVersionAsReceived(event *watchCacheEvent) {
+	if event.Type == watch.Bookmark {
+		c.stateMutex.Lock()
+		defer c.stateMutex.Unlock()
+		if c.wasBookmarkAfterResourceVersionReceivedLocked() {
+			return
+		}
+		// bookmark events are scheduled by startDispatchingBookmarkEvents method
+		// since we received a bookmark event that means we have
+		// converged towards the expected RV and it is okay to update the state so that
+		// this cacher can be scheduler for a regular bookmark events
+		c.state = cacheWatcherBookmarkReceived
+	}
+}
+
+// wasBookmarkAfterResourceVersionSent checks if a bookmark event with an RV >= the bookmarkAfterResourceVersion has been sent by this watcher
+func (c *cacheWatcher) wasBookmarkAfterResourceVersionSent() bool {
```

Comment:
After applying my comments from above - I suggest inlining the only remaining call to this function (in markBookmarkAfter.. and getting rid of this function.